### PR TITLE
Add PyTorch DataLoader Evaluator plugin

### PR DESCRIPTION
**Pull request by JanuszL**
- Introduces a lightweight diagnostic tool for identifying data loading
  bottlenecks in PyTorch training pipelines.
- This change adds Loader Evaluator inside pytorch DALI plugin, a
  jupyter notebook tutorial, and a documentation page with tests
- LoaderEvaluator class wraps PyTorch DataLoader with performance monitoring
  Two operation modes: 'log' (normal iteration with metrics) and 'replay'
  (cached batches for ideal performance simulation)
- PerformanceMetrics class for detailed performance tracking and bottleneck
  analysis
- In-memory batch caching for replay mode to simulate ideal data loading
- Comprehensive test suite and documentation with example notebook
- The tool helps users compare real vs. ideal data loading performance and
  identify optimization opportunities.

Authored-by: Albert Wolant <awolant@nvidia.com>

## Category:
**New feature** (*non-breaking change which adds functionality*)

## Description:
- Introduces a lightweight diagnostic tool for identifying data loading
  bottlenecks in PyTorch training pipelines.
- This change adds Loader Evaluator inside pytorch DALI plugin, a
  jupyter notebook tutorial, and a documentation page with tests
- LoaderEvaluator class wraps PyTorch DataLoader with performance monitoring
  Two operation modes: 'log' (normal iteration with metrics) and 'replay'
  (cached batches for ideal performance simulation)
- PerformanceMetrics class for detailed performance tracking and bottleneck
  analysis
- In-memory batch caching for replay mode to simulate ideal data loading
- Comprehensive test suite and documentation with example notebook
- The tool helps users compare real vs. ideal data loading performance and
  identify optimization opportunities.

### Affected modules and functionalities:
- new module in Pytorch plugin
- new example
- new test for it
- new documentation page describing the overall idea

### Key points relevant for the review:
- overall idea and flow

### Tests:

- [ ] Existing tests apply
- [x] New tests added
  - [x] Python tests
      - test_pytorch_loader_evaluator.py
  - [ ] GTests
  - [ ] Benchmark
  - [ ] Other
- [ ] N/A

### Documentation
- [ ] Existing documentation applies
- [x] Documentation updated
  - [ ] Docstring
  - [ ] Doxygen
  - [x] RST
      - pytorch_data_loader_evaluator.rst
  - [x] Jupyter
      - pytorch_data_loader_evaluator.ipynb
  - [ ] Other
- [ ] N/A

#### Requirements
- [ ] Implements new requirements
- [ ] Affects existing requirements
- [x] N/A

**REQ IDs**: N/A

**JIRA TASK**: DALI-4299

### CODE REVIEW

**Comment by JanuszL** [comment]
!build

File: qa/TL0_python-self-test-core/test_body.sh [hunk 1]
```diff
@@ -22,6 +22,7 @@ test_py_with_framework() {
     done
 
     ${python_new_invoke_test} -A '!slow,!pytorch,!mxnet,!cupy' test_backend_impl
+    ${python_new_invoke_test} -A '!slow,!pytorch,!mxnet,!cupy' test_pytorch_loader_evaluator
```
Comment (mdabek-nvidia) [inline qa/TL0_python-self-test-core/test_body.sh]: Is it necessary to run these tests in L0 maybe we should move it to lower level?

**Comment by JanuszL** [comment]
@greptileai please review again.

**Comment by JanuszL** [comment]
!build

**Comment by JanuszL** [comment]
@greptileai please review again.

**Comment by JanuszL** [comment]
@greptileai please review again.

**Comment by JanuszL** [comment]
!build

File: docs/plugins/pytorch_data_loader_evaluator.rst [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+   # 1. Logging real performance
+   wrapped_loader = LoaderEvaluator(dataloader, mode="log")
+   train_one_epoch(model, wrapped_loader, log_file="logged_real_run.json")
+
+   # 2. Simulating ideal performance (automatically caches batches during construction)
+   replay_loader = LoaderEvaluator(dataloader, mode="replay", num_cached_batches=20)
+   train_one_epoch(model, replay_loader, log_file="logged_ideal_run.json")
+
+   # 3. Visualizing results
+   plot_comparison(
```
Comment (mdabek-nvidia) [inline docs/plugins/pytorch_data_loader_evaluator.rst:69]: Shouldn't we add it to Jupyter Notebook as well?

File: docs/examples/frameworks/pytorch/loader_evaluator/pytorch_data_loader_evaluator.ipynb [hunk 3]
```diff
@@ -0,0 +1,451 @@
+{
```
Comment (mdabek-nvidia) [inline docs/examples/frameworks/pytorch/loader_evaluator/pytorch_data_loader_evaluator.ipynb:1]: <pre class="ql-syntax" spellcheck="false">Line #10.    import sys
</pre><p>I think that sys is not used anywhere.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/NVIDIA/DALI/pull/6112/discussion/#comment-2592958297'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='2'/>

Comment (mdabek-nvidia) [inline docs/examples/frameworks/pytorch/loader_evaluator/pytorch_data_loader_evaluator.ipynb:1]: <pre class="ql-syntax" spellcheck="false">Line #1.    # Wrap your dataloader with LoaderEvaluator (this is the only change!)
</pre><p>Nitpick: there is also a change below <code>dataloader</code> became <code>monitored_dataloader</code></p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/NVIDIA/DALI/pull/6112/discussion/#comment-2592958300'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='10'/>

File: qa/TL0_python-self-test-core/test_body.sh (hunk 1 above)
Comment (JanuszL) [reply qa/TL0_python-self-test-core/test_body.sh]: On my machine it is:
```
Ran 20 tests in 0.048s
```
So negligible addition if that concerns you.

File: docs/plugins/pytorch_data_loader_evaluator.rst (hunk 2 above)
Comment (JanuszL) [reply docs/plugins/pytorch_data_loader_evaluator.rst:69]: I think TL1_jupyter_plugins/test_pytorch.sh should cover this as it runs:
```
find frameworks/pytorch/ -name "*.ipynb" ...
```

File: docs/examples/frameworks/pytorch/loader_evaluator/pytorch_data_loader_evaluator.ipynb (hunk 3 above)
Comment (JanuszL) [reply docs/examples/frameworks/pytorch/loader_evaluator/pytorch_data_loader_evaluator.ipynb:1]: Do you think it should be:
```
dataloader = LoaderEvaluator(
    dataloader, mode="replay", num_cached_batches=len(dataloader) // 10
)
```
instead?

File: qa/TL0_python-self-test-core/test_body.sh (hunk 1 above)
Comment (mdabek-nvidia) [reply qa/TL0_python-self-test-core/test_body.sh]: Yes, this was my concern. If it is that fast, we can leave it.

File: docs/examples/frameworks/pytorch/loader_evaluator/pytorch_data_loader_evaluator.ipynb (hunk 3 above)
Comment (mdabek-nvidia) [reply docs/examples/frameworks/pytorch/loader_evaluator/pytorch_data_loader_evaluator.ipynb:1]: Yes, this would be clear, that we are just "updating" the dataloader.

Comment (JanuszL) [reply docs/examples/frameworks/pytorch/loader_evaluator/pytorch_data_loader_evaluator.ipynb:1]: Fixed

Comment (JanuszL) [reply docs/examples/frameworks/pytorch/loader_evaluator/pytorch_data_loader_evaluator.ipynb:1]: Fixed

**Comment by JanuszL** [comment]
!build

**Comment by JanuszL** [comment]
!build

File: dali/python/nvidia/dali/plugin/pytorch/loader_evaluator/loader.py [hunk 4]
```diff
@@ -0,0 +1,207 @@
...
+
+        self.start_time = time.time()
+        self.batch_times = []
+
+        # Replay cached batches to match original DataLoader length
+        original_length = len(self.dataloader)
+
+        for i in range(original_length):
+            batch_start = time.time()
+            batch = self.cached_batches[i % len(self.cached_batches)]
```
Comment (jantonguirao) [inline dali/python/nvidia/dali/plugin/pytorch/loader_evaluator/loader.py:142]: shouldn't you use `self.cache_lock` here as well, or is it guaranteed that only one thread can access the cached_batches structure once in replay mode?

File: dali/python/nvidia/dali/plugin/pytorch/loader_evaluator/loader.py [hunk 5]
```diff
@@ -0,0 +1,207 @@
...
+            batch_time = time.time() - batch_start
+            self.batch_times.append(batch_time)
+            yield batch
+
+        self.end_time = time.time()
+
+    def get_metrics(self) -> Dict[str, Any]:
+        """Return collected performance metrics."""
+        if not self.batch_times:
```
Comment (jantonguirao) [inline dali/python/nvidia/dali/plugin/pytorch/loader_evaluator/loader.py:151]: should we also check that `self.end_time` is set?

File: dali/test/python/test_pytorch_loader_evaluator.py [hunk 6]
```diff
@@ -0,0 +1,444 @@
...
+from nvidia.dali.plugin.pytorch.loader_evaluator import LoaderEvaluator  # noqa: E402
+
+
+class TestLoaderEvaluatorBasic(unittest.TestCase):
+    """Test basic functionality of LoaderEvaluator."""
+
+    def setUp(self):
+        """Create sample data for testing."""
+        # Create simple tensor data
+        data = torch.randn(100, 3, 32, 32)  # 100 samples, 3 channels, 32x32 images
```
Comment (jantonguirao) [inline dali/test/python/test_pytorch_loader_evaluator.py:36]: consider setting the torch seed somewhere, for reproducibility

File: dali/test/python/test_pytorch_loader_evaluator.py [hunk 7]
```diff
@@ -0,0 +1,444 @@
...
+        # Process some batches
+        batch_count = 0
+        for batch in loader:
+            batch_count += 1
+            if batch_count >= 2:
+                break
+
+        # Capture stdout
+        import sys
+        from io import StringIO
```
Comment (jantonguirao) [inline dali/test/python/test_pytorch_loader_evaluator.py]: maybe move this to the top of the file?

**Review by jantonguirao — APPROVED** [review]

File: dali/python/nvidia/dali/plugin/pytorch/loader_evaluator/loader.py (hunk 5 above)
Comment (JanuszL) [reply dali/python/nvidia/dali/plugin/pytorch/loader_evaluator/loader.py:151]: Based on the code it seems that if `batch_times` others won't be as well, but just in case we can add tests for `start_time` and `start_time` as well.

File: dali/python/nvidia/dali/plugin/pytorch/loader_evaluator/loader.py (hunk 4 above)
Comment (JanuszL) [reply dali/python/nvidia/dali/plugin/pytorch/loader_evaluator/loader.py:142]: Mhh, I'm not sure if we need this `self.cache_lock` at all. `__init__` happens in the main thread, while batches can be created concurrently they are returned in order from the original data loader.
Same applies to accessing the loader - iterators by definition are not thread safe and [the user's code should do the locking](https://docs.python.org/3/howto/free-threading-python.html#iterators).

File: dali/test/python/test_pytorch_loader_evaluator.py (hunk 7 above)
Comment (JanuszL) [reply dali/test/python/test_pytorch_loader_evaluator.py]: Done

File: dali/test/python/test_pytorch_loader_evaluator.py (hunk 6 above)
Comment (JanuszL) [reply dali/test/python/test_pytorch_loader_evaluator.py:36]: Done

**Comment by JanuszL** [comment]
!build

File: dali/python/nvidia/dali/plugin/pytorch/loader_evaluator/loader.py (hunk 5 above)
Comment (JanuszL) [reply dali/python/nvidia/dali/plugin/pytorch/loader_evaluator/loader.py:151]: https://github.com/NVIDIA/DALI/pull/6112#pullrequestreview-3558772460 is right and I missed the line below. When `self.end_time` is not set we just set `total_time=0` which is fine.

**Comment by JanuszL** [comment]
!build

File: dali/python/nvidia/dali/plugin/pytorch/loader_evaluator/loader.py (hunk 4 above)
Comment (jantonguirao) [reply dali/python/nvidia/dali/plugin/pytorch/loader_evaluator/loader.py:142]: Got it. Better removed then
